CASMCVT-221 Update to create system config using CDS (#4026)
Cray-HPE/docs-csm: install/Create_System_Configuration_Using_Cluster_Discovery_Service.md

# Create System Configuration Using Cluster Discovery Service

This stage walks the user through creating the configuration payload for the system using Cluster Discovery Service (CDS) to generate seed files and paddle file.

Run the following steps before starting any of the system configuration procedures.

1. (`pit#`) Make the `prep` and `cds` directory.

   ```bash
   mkdir -pv "${PITDATA}/prep/cds"
   ```

1. (`pit#`) Change into the `cds` directory.

   ```bash
   cd "${PITDATA}/prep/cds"
   ```

1. (`pit#`) Set `CVT_PATH` variable.

   ```bash
   CVT_PATH=/opt/src/cluster-config-verification-tool/
   ```

## Topics

> 

Reviewer: Is the user expected to follow these in order, as a procedure? The use of a numbered list implies that. But earlier on the page it says "Run the following steps before starting any of the system configuration procedures", which makes it sound like maybe the user would only be doing some of these.

If this is one procedure that is intended to be done in order, then the section above should just be added as a prerequisites section to the steps below.

Author: The added procedure was almost similar to the earlier procedure that was using the SHCD method. So the similar statements were used. This also can be added as a separate Prerequisite section before the actual procedures. 

To maintain the similar flow of the earlier content, I added it under the same statement. 

Please suggest if this needs to be changed.

   >   [CSM Overview](../introduction/csm_overview.md).
   > - To enable or disable audit logging, refer to [Audit Logs](../operations/security_and_authentication/Audit_Logs.md)
   >   for more information.
   > - If the system is using a `cabinets.yaml` file, be sure to update the `cabinets-yaml` field with `'cabinets.yaml'` as its value.

   ```bash
   vim system_config.yaml
   ```

## 2. Configure the management network

1. (`pit#`) Change into the `cds` directory.

   ```bash
   cd "${PITDATA}/prep/cds"
   ```

1. (`pit#`) Set up the `bond0` network interface as follows:

   - Configure `bond0` interface for single VLAN:

     ```bash
     "${CVT_PATH}"/cds/network_setup.py --mtlip <ipaddr>,<netmask>,<slave0>,<slave1>
     ```

   - Configure `bond0` interface for HMN configured network (VLAN 1, 2, and 4):

     ```bash
     "${CVT_PATH}"/cds/network_setup.py --mtlip <ipaddr>,<netmask>,<slave0>,<slave1> --hmnip <ipaddr>,<netmask>,<vlanid> --nmnip <ipaddr>,<netmask>,<vlanid>
     ```

     The variables and their descriptions are as follows:

     | Variable  | Description  |
     | ---       |  ---         |

>      | ipaddr    | IP address   |
>      | netmask   | netmask value |
>      | slave0, slave1 | network interface |
>      | vlanid    | VLAN ID number |

Reviewer: ```suggestion
     | `ipaddr`    | IP address   |
     | `netmask`   | netmask value |
     | `slave0`, `slave1` | network interfaces |
     | `vlanid`    | VLAN ID number |
```

     > **NOTE**: The IP address value, netmask value, network interface, and VLAN ID are available in the `system_config.yaml` file in the `prep` directory.

## 3. Configure auto-discovery

### 3.1 Configure Dnsmasq

1. (`pit#`) Configure Dnsmasq and PXE Boot.

   - Configure Dnsmasq for single VLAN:

> 
>      ```bash
>      "${CVT_PATH}"/cds/discover_enable.py --iprange <dhcp_ip_start>,<dhcp_ip_end>,<subnet>,<leasetime> --interface bond0
>      ```
> 
>      Where, `leasetime` is an optional variable.

Reviewer: ```suggestion
     > The `<leasetime>` parameter is optional.

     ```bash
     "${CVT_PATH}"/cds/discover_enable.py --iprange <dhcp_ip_start>,<dhcp_ip_end>,<subnet>,<leasetime> --interface bond0
     ```
```

     Example command:

     ```bash
     "${CVT_PATH}"/cds/discover_enable.py --iprange 10.x.x.x,10.x.x.x,255.x.x.x,24h --interface bond0
     ```

   - Configure Dnsmasq for HMN configured network (VLAN 1, 2, and 4):

     ```bash
     "${CVT_PATH}"/cds/discover_enable.py --mtlrange <admin_node_mtl_ip>,<start_ip>,<end_ip> \
                                          --nmnrange <admin_node_nmn_ip>,<start_ip>,<end_ip> \
                                          --hmnrange <admin_node_hmn_ip>,<start_ip>,<end_ip>
     ```

     Example command:

     ```bash
     "${CVT_PATH}"/cds/discover_enable.py --mtlrange 10.x.x.x,10.x.x.x,10.x.x.x \
                                          --nmnrange 10.x.x.x,10.x.x.x,10.x.x.x \
                                          --hmnrange 10.x.x.x,10..x.x.x,10.x.x.x
     ```

     > **NOTE**: The `iprange` for single VLAN or `mtliprange`, `nmniprange`, and `hmniprange` for HMN configured network are available in the `system_config.yaml` file in the `prep` directory.

### 3.2 Verify and reset the BMCs

> The following steps will verify if IP addresses are assigned to the BMC nodes and reset the nodes:

Reviewer: ```suggestion
The following steps will verify if IP addresses are assigned to the BMC nodes, and then reset the nodes.
```

1. (`pit#`) If the fabric switches, PDUs, and CMCs are to be dynamically discovered, then ensure that they are set to DHCP mode.

1. (`pit#`) List the BMC nodes.

   ```bash
   "${CVT_PATH}"/cds/discover_start.py -u bmc_username -p bmc_password --list
   ```

1. Verify that all the required nodes are listed.

   Example output:

   ![List the BMC nodes](../img/install/List_BMC_nodes.png)

1. (`pit#`) Reset the listed BMC nodes.

   ```bash
   "${CVT_PATH}"/cds/discover_start.py -u bmc_username -p bmc_password --reset
   ```

### 3.3 Verify the discover status

> **NOTE**: If the fabric switches, PDUs, and CMCs were not configured to DHCP mode in the step [Verify and reset the BMCs](#32-verify-and-reset-the-bmcs), manually create the `fabricswlist`, `pdulist`, and `cmclist` files in the working directory.

The following steps verify the status and lists the IP addresses of nodes, fabric switches, PDUs, and CMCs:

1. (`pit#`) Verify the status and list the IP addresses of the nodes.

   ```bash
   "${CVT_PATH}"/cds/discover_status.py nodes --out
   ```

>    1. (`pit#`) Create a `bond0` network interface on all the listed nodes.

Reviewer: Do not need the command prompt here if it is present in the parent step above.

```suggestion
   1. Create a `bond0` network interface on all the listed nodes.
```

   ```bash
   "${CVT_PATH}"/cds/discover_status.py fabric --username fabric_username --password fabric_password --out
   ```

1. (`pit#`) Verify the status and list the IP addresses of the PDUs.

   ```bash
   "${CVT_PATH}"/cds/discover_status.py pdu --username pdu_username --password pdu_password --out
   ```

1. (`pit#`) Verify the status and list the IP addresses of the CMCs.

   ```bash
   "${CVT_PATH}"/cds/discover_status.py cmc --username cmc_username --password cmc_password --out
   ```

### 3.4 Collect and compare the inventory data

#### 3.4.1 Dynamic Data Collection

1. (`pit#`) Copy the SSH key to setup passwordless-SSH for the local host.

   ```bash
   ssh-copy-id localhost
   ```

1. (`pit#`) Collect the hardware inventory data.

   ```bash
   "${CVT_PATH}"/system_inventory.py --nodefile nodelist
   ```

1. (`pit#`) Collect the fabric inventory data.

   ```bash
   "${CVT_PATH}"/fabric_inventory.py --switchfile ./cds/fabricswlist --username switch_username --password switch_password
   ```

1. (`pit#`) Collect the management network inventory.

   1. Manually, create a `mgmtswlist` file in the working directory listing the IP addresses of management switches.

>    2. (`pit#`) Collect the management network inventory data.

Reviewer: ```suggestion
   1. Collect the management network inventory data.
```

      ```bash
      "${CVT_PATH}"/management_inventory.py --switchfile mgmtswlist --password password --username username
      ```

1. (`pit#`) Collect the PDU inventory data.

   ```bash
   "${CVT_PATH}"/cds/pdu_cmc_inventory.py --pdu_file pdulist
   ```

1. (`pit#`) Collect the CMC inventory data.

   ```bash
   "${CVT_PATH}"/cds/pdu_cmc_inventory.py --cmc_file cmclist
   ```

#### 3.4.2 SHCD data population

1. Copy the SHCD files to the present working directory.

1. (`pit#`) Create a JSON file using the CANU tool.

   See [Validate the SHCD](../operations/network/management_network/validate_shcd.md#validate-the-shcd).

   Example command:

   ```bash
   canu validate shcd --shcd SHCD.xlsx --architecture V1 --tabs 40G_10G,NMN,HMN --corners I12,Q38,I13,Q21,J20,U38 --json --out cabling.json
   ```

1. (`pit#`) Store the SHCD data in the CVT database.

   ```bash
   "${CVT_PATH}"/parse_shcd.py --canu_json_file cabling.json
   ```

#### 3.4.3 Server classification

> 1. (`pit#`) Classify the servers.
> 
>    > In the following command, replace `<N>` with the number of worker NCNs in the system.
> 
>    ```bash
>    "${CVT_PATH}"/scripts/server_classification.py --no_of_workers <N>
>    ```

Reviewer: Sections with a single step should not have that step enumerated.
````suggestion
(`pit#`) Classify the servers.

> In the following command, replace `<N>` with the number of worker NCNs in the system.

```bash
"${CVT_PATH}"/scripts/server_classification.py --no_of_workers <N>
```
````

#### 3.4.4 Compare the SHCD data with CVT inventory data

1. (`pit#`) List the snapshot IDs of the SHCD and the CVT inventory data.

   ```bash
   "${CVT_PATH}"/shcd_compare.py --list
   ```

1. (`pit#`) Compare the SHCD data with the CVT inventory data.

   ```bash
   "${CVT_PATH}"/shcd_compare.py --shcd_id <SnapshotID_SHCD> --cvt_id <SnapshotID_CVT>
   ```

   Example command:

   ```bash
   "${CVT_PATH}"/shcd_compare.py --shcd_id 4ea86f99-47ca-4cd9-b142-ea572d0566bf --cvt_id 326a6edd-adce-4d1f-9f30-ae9539284733
   ```

## 4. Generate seed files and paddle file

1. (`pit#`) Generate the following seed files: `switch_metadata.csv`, `application_node_config.yaml`, `hmn_connections.json`, and `ncn_metadata.csv`.

   ```bash
   "${CVT_PATH}"/create_seedfile.py
   ```

   > **NOTE**: The generated seed files will be stored in the `seedfiles` folder in the working directory.

1. (`pit#`) Generate paddle file: `cvt-ccj.json`.

>    ```bash
>    "${CVT_PATH}"/create_paddle_file.py --architecture <arch_type>
>    ```
> 
>    Where, the variable `<arch_type>` describes the type of architecture (FULL, V1, or TDS).
>    See [Generate configuration files](../operations/network/management_network/generate_switch_configs.md#generate-configuration-files) for the characteristics of the architecture.

Reviewer: ````suggestion
   > In the following command, replace `<arch_type>` with the type of architecture (`Full`, `V1`, or `Tds`).
   > See [Generate configuration files](../operations/network/management_network/generate_switch_configs.md#generate-configuration-files) for the characteristics of the architecture.

   ```bash
   "${CVT_PATH}"/create_paddle_file.py --architecture <arch_type>
   ```
````

1. Copy the generated paddle file and seed files to the `prep` directory.

   ```bash

>    cp -r ${PITDATA}/prep/cds/cvt-ccj.json ${PITDATA}/prep/cds/seedfiles/* ${PITDATA}/prep/

Reviewer: ```suggestion
   cp -rv ${PITDATA}/prep/cds/cvt-ccj.json ${PITDATA}/prep/cds/seedfiles/* ${PITDATA}/prep/
```

   ```

1. Create `cabinets.yaml` file in the `prep` directory, see [Create `cabinets.yaml`](create_cabinets_yaml.md).

## 5. Stop discovery

> 1. (`pit#`) Clean and reset the modified files to their original default state.
> 
>    ```bash
>    "${CVT_PATH}"/cds/discover_stop.py -u bmc_username -p bmc_password
>    ```
> 
> 1. (`pit#`) Change into the `prep` directory.
> 
>    ```bash
>    cd "${PITDATA}/prep"
>    ```
> 
> ## 6. Run CSI
> 
> 1. (`pit#`) Generate the initial configuration for CSI.

Reviewer: The change directory step makes more sense as the first step in section 6 than as the final step in section 5.
````suggestion
(`pit#`) Clean and reset the modified files to their original default state.

```bash
"${CVT_PATH}"/cds/discover_stop.py -u bmc_username -p bmc_password
```

## 6. Run CSI

1. (`pit#`) Change into the `prep` directory.

   ```bash
   cd "${PITDATA}/prep"
   ```

1. (`pit#`) Generate the initial configuration for CSI.
````
